### PR TITLE
Solve Day 3: Binary Diagnostic puzzle

### PR DESCRIPTION
**Pull request by roman-khomenko-dev**
Implement solution for the Day 3 puzzle

### CODE REVIEW

File: lib/day_3/binary_diagnostic.ex [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+      |> List.first()
+      |> String.length()
+
+    with oxygen <- get_rating_value(width, {data, "common"}),
+         scrubber <- get_rating_value(width, {%{puzzle: elem(get_puzzle(), 1)}, "less"}) do
+      oxygen * scrubber
+    end
+  end
+
+  defp get_rating_value(width, {data, type}) when type in ["common", "less"] do
```
Comment (sergey-lukianov) [inline lib/day_3/binary_diagnostic.ex]: why to use strings?

File: lib/day_3/binary_diagnostic.ex [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    |> Enum.map(&get_index_diff(puzzle, &1))
+    |> Enum.reduce(data, fn row, acc ->
+      %{
+        acc
+        | gamma_rate: acc.gamma_rate <> common_bit(row),
+          epsilon_rate: acc.epsilon_rate <> less_bit(row)
+      }
+    end)
+    |> Enum.into([], fn {_key, value} -> get_decimal(value) end)
+    |> Enum.product()
```
Comment (sergey-lukianov) [inline lib/day_3/binary_diagnostic.ex]: write obvious comments of what you doing in terms of the task

File: lib/day_3/binary_diagnostic.ex [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+      zero: get_index_count(puzzle, index, "0"),
+      one: get_index_count(puzzle, index, "1")
+    }
+  end
+
+  defp get_decimal(number) when is_binary(number), do: elem(Integer.parse(number, 2), 0)
+
+  defp get_index_count(puzzle, index, bit), do: Enum.count(puzzle, &(String.at(&1, index) == bit))
+
+  defp common_bit(%{one: one, zero: zero} = _row), do: if(one >= zero, do: "1", else: "0")
```
Comment (sergey-lukianov) [inline lib/day_3/binary_diagnostic.ex]: `dominant_bit_value` would be a better name

File: lib/day_3/binary_diagnostic.ex [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+  defp less_bit(%{one: one, zero: zero} = _row), do: if(zero <= one, do: "0", else: "1")
```
Comment (sergey-lukianov) [inline lib/day_3/binary_diagnostic.ex]: come up with another name

File: lib/day_3/binary_diagnostic.ex [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+  defp get_index_diff(puzzle, index) do
+    %{
+      index: index,
+      zero: get_index_count(puzzle, index, "0"),
+      one: get_index_count(puzzle, index, "1")
+    }
+  end
+
+  defp get_decimal(number) when is_binary(number), do: elem(Integer.parse(number, 2), 0)
```
Comment (sergey-lukianov) [inline lib/day_3/binary_diagnostic.ex]: rewrite with pipe

File: lib/day_3/binary_diagnostic.ex (hunk 1 above)
Comment (roman-khomenko-dev) [reply lib/day_3/binary_diagnostic.ex]: changed to atoms

File: lib/day_3/binary_diagnostic.ex [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+  @spec power_consumption :: number
+  def power_consumption do
+    {_, puzzle} = get_puzzle()
+    data = %{gamma_rate: "", epsilon_rate: ""}
+
+    # Get width of one string row to make calculations on each of them
+    width =
+      puzzle
+      |> List.first()
+      |> String.length()
```
Comment (sergey-lukianov) [inline lib/day_3/binary_diagnostic.ex]: so make this block  a function

and name variables to reflect whats happening

File: lib/day_3/binary_diagnostic.ex [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+    # Get map with index and count values to compare
+    0..(width - 1)
+    |> Enum.map(&get_index_diff(puzzle, &1))
```
Comment (sergey-lukianov) [inline lib/day_3/binary_diagnostic.ex]: reimplementing `Enum.zip_index`?

File: lib/day_3/binary_diagnostic.ex [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    # Compose gamma and epsilon strings
+    |> Enum.reduce(data, fn row, acc ->
+      %{
+        acc
+        | gamma_rate: acc.gamma_rate <> dominant_bit_value(row),
+          epsilon_rate: acc.epsilon_rate <> least_bit_value(row)
+      }
+    end)
```
Comment (sergey-lukianov) [inline lib/day_3/binary_diagnostic.ex]: make a funciton

File: lib/day_3/binary_diagnostic.ex [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+    # Put values to list and calculate multiply result
+    |> Enum.into([], fn {_key, value} -> get_decimal(value) end)
+    |> Enum.product()
```
Comment (sergey-lukianov) [inline lib/day_3/binary_diagnostic.ex]: make a function

File: lib/day_3/binary_diagnostic.ex [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+  end
+
+  @spec life_support_rating :: integer
+  def life_support_rating do
+    data = %{puzzle: elem(get_puzzle(), 1)}
+
+    width =
+      data.puzzle
+      |> List.first()
+      |> String.length()
```
Comment (sergey-lukianov) [inline lib/day_3/binary_diagnostic.ex]: copy-paste?

File: lib/day_3/binary_diagnostic.ex [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+        acc
+        | puzzle: Enum.filter(acc.puzzle, fn row -> String.at(row, index) == common end)
+      }
+
+      if length(acc.puzzle) > 1, do: {:cont, acc}, else: {:halt, acc}
+    end)
+    |> Map.get(:puzzle)
+    |> List.first()
+    |> get_decimal()
+  end
```
Comment (sergey-lukianov) [inline lib/day_3/binary_diagnostic.ex]: make some comments

File: lib/day_3/binary_diagnostic.ex (hunk 7 above)
Comment (roman-khomenko-dev) [reply lib/day_3/binary_diagnostic.ex]: yes, but not need additional indexing actually here so it removed

File: lib/day_3/binary_diagnostic.ex (hunk 10 above)
Comment (roman-khomenko-dev) [reply lib/day_3/binary_diagnostic.ex]: change with `puzzle_symbol_range` function

File: lib/day_3/binary_diagnostic.ex (hunk 11 above)
Comment (roman-khomenko-dev) [reply lib/day_3/binary_diagnostic.ex]: commented

File: lib/day_3/binary_diagnostic.ex (hunk 9 above)
Comment (roman-khomenko-dev) [reply lib/day_3/binary_diagnostic.ex]: add `calculate_result/1` function

File: lib/day_3/binary_diagnostic.ex (hunk 8 above)
Comment (roman-khomenko-dev) [reply lib/day_3/binary_diagnostic.ex]: add `compose_characteristics/1` function

File: lib/day_3/binary_diagnostic.ex (hunk 6 above)
Comment (roman-khomenko-dev) [reply lib/day_3/binary_diagnostic.ex]: moved to `defp puzzle_symbol_range/1`

File: lib/day_3/binary_diagnostic.ex [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+    |> compose_characteristics(data)
+    # Put values to list and calculate multiply result
+    |> calculate_result()
+  end
+
+  @spec life_support_rating :: integer
+  def life_support_rating do
+    data = %{puzzle: elem(get_puzzle(), 1)}
+
+    with oxygen <- get_rating_value(data.puzzle, {data, :common}),
```
Comment (sergey-lukianov) [inline lib/day_3/binary_diagnostic.ex]: rename `:common` and `:less` throughout the code

File: lib/day_3/binary_diagnostic.ex [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+  end
+
+  defp extract_result(data) do
+    data
+    |> Map.get(:puzzle)
+    |> List.first()
+    |> get_decimal()
+  end
+
+  defp calculate_result(data) do
```
Comment (sergey-lukianov) [inline lib/day_3/binary_diagnostic.ex]: Too abstract naming, rename in more descriptive, like
- string_list_to_product
- calculate_product_of_numbers

File: lib/day_3/binary_diagnostic.ex [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+
+      if length(acc.puzzle) > 1, do: {:cont, acc}, else: {:halt, acc}
+    end)
+  end
+
+  defp apply_criteria_filter(bit, {acc, index}) do
+    %{acc | puzzle: Enum.filter(acc.puzzle, fn row -> String.at(row, index) == bit end)}
+  end
+
+  defp extract_result(data) do
```
Comment (sergey-lukianov) [inline lib/day_3/binary_diagnostic.ex]: Too abstract name, ideally should reflect what is going here

File: lib/day_3/binary_diagnostic.ex (hunk 12 above)
Comment (roman-khomenko-dev) [reply lib/day_3/binary_diagnostic.ex]: those names changed to :dominant, :least

File: lib/day_3/binary_diagnostic.ex (hunk 13 above)
Comment (roman-khomenko-dev) [reply lib/day_3/binary_diagnostic.ex]: renamed to `calculate_power_consumption`

File: lib/day_3/binary_diagnostic.ex (hunk 14 above)
Comment (roman-khomenko-dev) [reply lib/day_3/binary_diagnostic.ex]: renamed as `extract_characteristic_rating`

**Review by sergey-lukianov — APPROVED** [review]

**Pull request merged**